File: frontend/src/api/project.ts
import { client } from './client'
import type { Application, ContainerRegistry, KubernetesCluster, Project, ProjectMember } from '../types'

export interface ProjectPayload {
  key?: string
  name?: string
  description?: string
  status?: string
  business_owner?: string
  billing_owner?: string
  github_group?: string
  github_default_visibility?: string
  aliyun_account_id?: string
  aliyun_resource_group_id?: string
  aliyun_region?: string
  aliyun_vpc_id?: string
  aliyun_binding_status?: string
  owner_name?: string
  owner_email?: string
  owner_title?: string
}

export const projectApi = {
  list: () => client.get<never, Project[]>('/projects'),
  get: (id: number) => client.get<never, Project>(`/projects/${id}`),
  create: (input: ProjectPayload & { key: string; name: string }) =>
    client.post<never, Project>('/projects', input),
  update: (id: number, input: ProjectPayload) =>
    client.patch<never, Project>(`/projects/${id}`, input),
  remove: (id: number) => client.delete(`/projects/${id}`),
  members: (projectId: number) =>
    client.get<never, ProjectMember[]>(`/projects/${projectId}/members`),
  addMember: (projectId: number, input: Partial<ProjectMember>) =>
    client.post<never, ProjectMember>(`/projects/${projectId}/members`, input),
  updateMember: (projectId: number, memberId: number, input: Partial<ProjectMember>) =>
    client.patch<never, ProjectMember>(`/projects/${projectId}/members/${memberId}`, input),
  removeMember: (projectId: number, memberId: number) =>
    client.delete(`/projects/${projectId}/members/${memberId}`),
  clusters: (projectId: number) =>
    client.get<never, KubernetesCluster[]>(`/projects/${projectId}/clusters`),
  addCluster: (projectId: number, input: Partial<KubernetesCluster>) =>
    client.post<never, KubernetesCluster>(`/projects/${projectId}/clusters`, input),
  updateCluster: (projectId: number, clusterId: number, input: Partial<KubernetesCluster>) =>
    client.patch<never, KubernetesCluster>(`/projects/${projectId}/clusters/${clusterId}`, input),
  removeCluster: (projectId: number, clusterId: number) =>
    client.delete(`/projects/${projectId}/clusters/${clusterId}`),
  setDefaultCluster: (projectId: number, clusterId: number) =>
    client.post<never, KubernetesCluster>(`/projects/${projectId}/clusters/${clusterId}/default`),
  applications: (projectId: number) =>
    client.get<never, Application[]>('/applications', { params: { projectId } }),
  registries: (projectId: number) =>
    client.get<never, ContainerRegistry[]>('/registries', { params: { projectId } }),
}

File: backend/app/models/project.py
from datetime import datetime, timezone

from app.extensions import db


def utcnow():
    return datetime.now(timezone.utc)


class Project(db.Model):
    __tablename__ = "projects"

    id = db.Column(db.Integer, primary_key=True)
    key = db.Column(db.String(64), nullable=False, unique=True, index=True)
    name = db.Column(db.String(120), nullable=False)
    description = db.Column(db.Text)
    status = db.Column(db.String(30), default="active", nullable=False, index=True)
    business_owner = db.Column(db.String(120))
    billing_owner = db.Column(db.String(120))
    github_group = db.Column(db.String(255))
    github_default_visibility = db.Column(db.String(30), default="private", nullable=False)
    aliyun_account_id = db.Column(db.String(64))
    aliyun_resource_group_id = db.Column(db.String(120))
    aliyun_region = db.Column(db.String(64))
    aliyun_vpc_id = db.Column(db.String(120))
    aliyun_binding_status = db.Column(
        db.String(30), default="unbound", nullable=False, index=True
    )
    created_at = db.Column(db.DateTime(timezone=True), default=utcnow, nullable=False)
    updated_at = db.Column(
        db.DateTime(timezone=True), default=utcnow, onupdate=utcnow, nullable=False
    )
    applications = db.relationship("Application", back_populates="project")
    members = db.relationship(
        "ProjectMember",
        back_populates="project",
        cascade="all, delete-orphan",
        order_by="ProjectMember.id",
    )
    kubernetes_clusters = db.relationship(
        "KubernetesCluster",
        back_populates="project",
        cascade="all, delete-orphan",
        order_by="KubernetesCluster.id",
    )
    registries = db.relationship(
        "ContainerRegistry",
        back_populates="project",
        cascade="all, delete-orphan",
        order_by="ContainerRegistry.name",
    )

    def to_dict(self, include_stats=False):
        data = {
            "id": self.id,
            "key": self.key,
            "name": self.name,
            "description": self.description,
            "status": self.status,
            "business_owner": self.business_owner,
            "billing_owner": self.billing_owner,
            "github_group": self.github_group,
            "github_default_visibility": self.github_default_visibility,
            "aliyun_account_id": self.aliyun_account_id,
            "aliyun_resource_group_id": self.aliyun_resource_group_id,
            "aliyun_region": self.aliyun_region,
            "aliyun_vpc_id": self.aliyun_vpc_id,
            "aliyun_binding_status": self.aliyun_binding_status,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
        }
        if include_stats:
            data.update({
                "applications_count": len(self.applications),
                "members_count": len(self.members),
                "clusters_count": len(self.kubernetes_clusters),
                "registries_count": len(self.registries),
            })
        return data
